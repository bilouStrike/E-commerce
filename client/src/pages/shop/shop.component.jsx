import React, { useEffect } from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';

import collectionsOverviewContainer from '../../components/collections-overview/collections-overview.container';
import collectionsContainer from '../../components/collection/collection.container';

import { fetchCollectionsStart } from './../../redux/shop/shop.actions';

export const ShopPage = ({ match, fetchCollectionsStart }) => {
  React.useEffect(() => {
    fetchCollectionsStart();
  }, [fetchCollectionsStart]);

  return (
    <div className="shop-page">
      <Route
        exact
        path={`${match.path}`}
        component={collectionsOverviewContainer}
      />
      <Route
        path={`${match.path}/:collectionId`}
        component={collectionsContainer}
      />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  fetchCollectionsStart: () => dispatch(fetchCollectionsStart()),
});

export default connect(null, mapDispatchToProps)(ShopPage);
