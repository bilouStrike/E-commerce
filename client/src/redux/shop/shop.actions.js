import shopActionsTypes from  './shop.types';

export const fetchCollectionsStart = () => ({
    type: shopActionsTypes.FETCH_COLLECTIONS_START
});

export const fetchCollectionsSuccess = (collectionMap) => ({
  type: shopActionsTypes.FETCH_COLLECTIONS_SUCCESS,
  payload: collectionMap
})

export const fetchCollectionsFailure = (errorMessage) => ({
  type: fetchCollectionsFailure,
  payload:errorMessage
})

