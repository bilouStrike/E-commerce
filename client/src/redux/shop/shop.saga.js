import { takeLatest, call, put }  from 'redux-saga/effects';
import shopActionsTypes from './shop.types';

import { firestore, converCollectionsSnapshotToMap } from '../../utility/firebase/firebase.utils';
import {
    fetchCollectionsSuccess,
    fetchCollectionsFailure
} from './../shop/shop.actions';

function* fetchCollectionAsync() {
    try {
        const collectionRef = yield firestore.collection('collections');
        const snapshot = yield collectionRef.get();
        const collectionMap = yield call(converCollectionsSnapshotToMap, snapshot);
        yield put(fetchCollectionsSuccess(collectionMap));
    } catch (error) {
        yield put(fetchCollectionsFailure(error.message));
    }
}

export default function* fetchCollectionsStart() {
    yield takeLatest(
        shopActionsTypes.FETCH_COLLECTIONS_START,
        fetchCollectionAsync
    )
}