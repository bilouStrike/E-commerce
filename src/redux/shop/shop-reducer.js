import shopActionsTypes from './shop.types';

const INITIAL_STATE = {};

const shopReducer = (state = INITIAL_STATE, action ) => {
  switch (action.type) {
    case shopActionsTypes.UPDATE_COLLECTION:
      return {
          ...state,
          collections:action.payload
      }
    default:
      return state
  }
}

export default shopReducer;