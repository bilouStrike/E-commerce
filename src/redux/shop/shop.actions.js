import shopActionsTypes from  './shop.types';

export const updateCollections = collectionsMap => ({
    type: shopActionsTypes.UPDATE_COLLECTION,
    payload: collectionsMap
  });