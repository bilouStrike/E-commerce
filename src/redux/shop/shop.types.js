const shopActionsTypes = {
    UPDATE_COLLECTION: 'UPDATE_COLLECTION'
}
export default shopActionsTypes;