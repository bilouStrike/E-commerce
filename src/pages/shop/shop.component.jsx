import React from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';

import CollectionsOverview from '../../components/collections-overview/collections-overview.component';
import CollectionPage from '../../components/collection/collection.component';
import { firestore, converCollectionsSnapshotToMap } from '../../utility/firebase/firebase.utils';
import { updateCollections } from './../../redux/shop/shop.actions';

class ShopPage extends React.Component {

  unsubscribeFromSnapshot = null;

  componentDidMount() {
      const {updateCollections} = this.props;
      const collectionRef = firestore.collection('collections');
      this.unsubscribeFromSnapshot = collectionRef.onSnapshot( async snapshot => {
          const collectionMap = converCollectionsSnapshotToMap(snapshot);
          console.log(collectionMap);
          updateCollections(collectionMap);
        }
      );
  }
  render(){
      const {match} = this.props;
      return (
          <div className='shop-page'>
            <Route exact path={`${match.path}`} component={CollectionsOverview} />
            <Route path={`${match.path}/:collectionId`} component={CollectionPage} />
          </div>
      );
  }
}
const mapDispatchToProps = dispatch => ({
  updateCollections: collectionMap => dispatch(updateCollections(collectionMap))
})
export default connect(null, mapDispatchToProps)(ShopPage);