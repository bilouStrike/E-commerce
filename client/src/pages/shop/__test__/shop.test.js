import { shallow } from 'enzyme';
import React from 'react';
import { ShopPage } from '../shop.component';
import toJson from 'enzyme-to-json';

describe("Shpe page", () => {
 
    let wrapper;
    let props;
    let useEffect;

    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    };
    beforeEach( () => { 
        useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
        props = {
            match : { 
                params: {id: 1},
                isExact: true,
                path: "",
                url: ""
            },
            fetchCollectionsStart: jest.fn().mockResolvedValue({ type: 'FETCH_COLLECTIONS_START'}),
        };
        mockUseEffect();
        mockUseEffect();
        wrapper = toJson(shallow(<ShopPage {...props} />));
    });

    describe("Shpe page", () => {
        it('expect to render shop page', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    it('expect to call callback on shallow shop page component', () => {
        expect(props.fetchCollectionsStart).toHaveBeenCalled();
    });
});