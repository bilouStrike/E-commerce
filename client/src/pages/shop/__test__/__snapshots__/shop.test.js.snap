// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shpe page Shpe page expect to render shop page 1`] = `
<div
  className="shop-page"
>
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(spinner)",
        "type": [Function],
      }
    }
    exact={true}
    path=""
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(spinner)",
        "type": [Function],
      }
    }
    path="/:collectionId"
  />
</div>
`;
